File: cpp/bench.cpp
#include <bits/stdc++.h>
#include <sys/resource.h>
#include <sys/time.h>
#include <unistd.h>

struct Result {
  std::string runtime, model;
  int threads, iterations, keys;
  double read_ratio;
  uint64_t seed;
  long long duration_ms;
  unsigned long long rss_bytes;
};

uint64_t splitmix64(uint64_t &x){
  x += 0x9E3779B97f4A7C15ULL;
  uint64_t z = x;
  z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9ULL;
  z = (z ^ (z >> 27)) * 0x94D049BB133111EBULL;
  z = z ^ (z >> 31);
  return z;
}

unsigned long long rss_bytes() {
  long rss=0L;
  FILE* fp = fopen("/proc/self/statm","r");
  if (!fp) return 0;
  if (fscanf(fp,"%*s%ld",&rss)!=1) { fclose(fp); return 0; }
  fclose(fp);
  return (unsigned long long)rss * sysconf(_SC_PAGESIZE);
}

int main(int argc, char** argv) {
  int threads=8, iterations=2'000'000, keys=100'000, shards=64;
  double read_ratio=0.9;
  uint64_t seed=42;
  for (int i=1;i<argc;i++) {
    std::string a=argv[i];
    if (a=="--threads") threads=std::stoi(argv[++i]);
    else if (a=="--iterations") iterations=std::stoi(argv[++i]);
    else if (a=="--keys") keys=std::stoi(argv[++i]);
    else if (a=="--read-ratio") read_ratio=std::stod(argv[++i]);
    else if (a=="--seed") seed=std::stoull(argv[++i]);
    else if (a=="--shards") shards=std::stoi(argv[++i]);
  }
  std::vector<std::unordered_map<int,long long>> maps(shards);
  std::vector<std::mutex> locks(shards);
  for (int i=0;i<keys;i++) maps[i%shards][i]=0;
  int per = iterations / threads;

  auto start = std::chrono::steady_clock::now();
  std::vector<std::thread> ts;
  ts.reserve(threads);
  for (int t=0;t<threads;t++) {
    ts.emplace_back([&, t](){
      uint64_t x = seed + t;
      int reads = int(read_ratio*1000);
      for (int i=0;i<per;i++) {
        int k = int(splitmix64(x) % uint64_t(keys));
        int r = int(splitmix64(x) % 1000ULL);
        int s = k % shards;
        if (r < reads) {
          std::lock_guard<std::mutex> g(locks[s]);
          auto it = maps[s].find(k);
          if (it!=maps[s].end()) (void)it->second;
        } else {
          std::lock_guard<std::mutex> g(locks[s]);
          maps[s][k] += 1;
        }
      }
    });
  }
  for (auto &th: ts) th.join();
  auto dur = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now()-start).count();
  printf("{\"runtime\":\"gcc\",\"model\":\"threads-sharded\",\"threads\":%d,\"iterations\":%d,\"keys\":%d,\"read_ratio\":%.3f,\"seed\":%llu,\"duration_ms\":%lld,\"rss_bytes\":%llu}\n",
    threads, per*threads, keys, read_ratio, (unsigned long long)seed, (long long)dur, rss_bytes());
  return 0;
}

File: rust/bench.rs
extern crate serde;
use serde::Serialize;
use std::{
    sync::{Arc, Mutex},
    time::Instant,
};

#[derive(Serialize)]
struct ResultOut {
    runtime: String,
    model: String,
    threads: usize,
    iterations: usize,
    keys: usize,
    read_ratio: f64,
    seed: u64,
    duration_ms: u128,
    rss_bytes: u64,
}

fn rss_bytes() -> u64 {
    #[cfg(target_os = "linux")]
    {
        let s = std::fs::read_to_string("/proc/self/statm").unwrap_or_default();
        let parts: Vec<&str> = s.split_whitespace().collect();
        if parts.len() >= 2 {
            let pages: u64 = parts[1].parse().unwrap_or(0);
            return pages * 4096;
        }
    }
    0
}

fn splitmix64(mut x: u64) -> impl FnMut() -> u64 {
    move || {
        x = x.wrapping_add(0x9E3779B97f4A7C15);
        let mut z = x;
        z = (z ^ (z >> 30)).wrapping_mul(0xBF58476D1CE4E5B9);
        z = (z ^ (z >> 27)).wrapping_mul(0x94D049BB133111EB);
        z ^ (z >> 31)
    }
}

fn run_sharded(
    threads: usize,
    iterations: usize,
    keys: usize,
    read_ratio: f64,
    seed: u64,
    shards_n: usize,
) {
    let mut shards = Vec::with_capacity(shards_n);
    for _ in 0..shards_n {
        shards.push(Arc::new(Mutex::new(vec![0i64; 0])));
    }
    // store as Vec of hashmaps but for speed, use Vec with capacity keys/shard and sparse fill via map-like approach
    let mut maps: Vec<Arc<Mutex<std::collections::HashMap<usize, i64>>>> =
        Vec::with_capacity(shards_n);
    for _ in 0..shards_n {
        maps.push(Arc::new(Mutex::new(
            std::collections::HashMap::with_capacity(keys / shards_n + 1),
        )));
    }
    for i in 0..keys {
        let s = &maps[i % shards_n];
        let mut g = s.lock().unwrap();
        g.insert(i, 0);
    }
    let per = iterations / threads;
    let start = Instant::now();
    let mut handles = Vec::new();
    for t in 0..threads {
        let maps = maps.clone();
        let reads = (read_ratio * 1000.0) as u64;
        let mut rnd = splitmix64(seed + t as u64);
        handles.push(std::thread::spawn(move || {
            for _ in 0..per {
                let k = (rnd() % keys as u64) as usize;
                let r = rnd() % 1000;
                let sidx = k % maps.len();
                let m = &maps[sidx];
                if r < reads {
                    let g = m.lock().unwrap();
                    let _ = g.get(&k);
                } else {
                    let mut g = m.lock().unwrap();
                    let e = g.entry(k).or_insert(0);
                    *e += 1;
                }
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }
    let dur = start.elapsed().as_millis();
    let out = ResultOut {
        runtime: format!("rustc{}", rustc_version_runtime()),
        model: "threads-sharded".to_string(),
        threads,
        iterations: per * threads,
        keys,
        read_ratio,
        seed,
        duration_ms: dur,
        rss_bytes: rss_bytes(),
    };
    println!("{}", serde_json::to_string(&out).unwrap());
}

fn rustc_version_runtime() -> String {
    // Best effort: read env var set by cargo; else unknown
    option_env!("RUSTC_VERSION")
        .unwrap_or("unknown")
        .to_string()
}

fn main() {
    let mut threads = 8usize;
    let mut iterations = 2_000_000usize;
    let mut keys = 100_000usize;
    let mut read_ratio = 0.9f64;
    let mut seed = 42u64;
    let mut _model = "threads".to_string();
    let mut shards = 64usize;
    let args: Vec<String> = std::env::args().collect();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--threads" => {
                i += 1;
                threads = args[i].parse().unwrap();
            }
            "--iterations" => {
                i += 1;
                iterations = args[i].parse().unwrap();
            }
            "--keys" => {
                i += 1;
                keys = args[i].parse().unwrap();
            }
            "--read-ratio" => {
                i += 1;
                read_ratio = args[i].parse().unwrap();
            }
            "--seed" => {
                i += 1;
                seed = args[i].parse().unwrap();
            }
            "--model" => {
                i += 1;
                _model = args[i].clone();
            }
            "--shards" => {
                i += 1;
                shards = args[i].parse().unwrap();
            }
            _ => {}
        }
        i += 1;
    }
    // For fairness use sharded Mutex<HashMap>; you can switch to DashMap by feature
    run_sharded(threads, iterations, keys, read_ratio, seed, shards);
}
